### PR TITLE
[OptApp] Add filtering factory

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR adds filtering factory which can be used among any controller. At the moment, this only supports explicit filters. The support for implicit filters will be done in a seperate PR.

This also fixes some of the problems in the filter functions. Following figure shows the fixed filter kerenels.
image

**🆕 Changelog**
- Update explicit filters
- Introduces the Filter::Factory
- Added tests

### CODE REVIEW

File: applications/OptimizationApplication/tests/test_OptimizationApplication.py [hunk 1]
```diff
@@ -118,7 +120,8 @@ def AssembleTestSuites():
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([algorithm_tests.analysis_based_tests.algorithm_gradient_projection.test_gradient_projection.TestGradientProjectionAnalysis]))
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([algorithm_tests.analysis_based_tests.algorithm_nesterov_accelerated_gradient.test_nestervo_accelerated_gradient_analysis.TestNesterovAcceleratedGradientAnalysis]))
 
-
+    # overall tests
+    smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([archara_shape_opt.test_archara_shape_opt.TestArcharaShapeOpt]))
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/test_OptimizationApplication.py]: please check if you have added your tests.

File: applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py [hunk 2]
```diff
@@ -79,5 +63,87 @@ def __RunConsistencyTest(self, vm_filter, entities, model_part):
...
+
+    def test_FilterLinear(self):
+        self.__RunTestCase("linear", "cosine", "explicit_filter_reference_linear.vtu.orig")
+
+    def test_FilterGaussian(self):
+        self.__RunTestCase("gaussian", "cosine", "explicit_filter_reference_gaussian.vtu.orig")
+
+    def test_FilterQuartic(self):
+        self.__RunTestCase("quartic", "cosine", "explicit_filter_reference_quartic.vtu.orig")
+
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py:154]: nice

File: applications/OptimizationApplication/python_scripts/filtering/explicit_vertex_morphing.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    return ExplicitVertexMorphingFilter(model, filtering_model_part_name, filtering_variable, data_location, parameters)
+
+class ExplicitVertexMorphingFilter(Filter):
+    @classmethod
+    def GetDefaultParameters(cls) -> Kratos.Parameters:
+        return Kratos.Parameters("""{
+            "filter_type"               : "explicit_vertex_morphing",
+            "filter_function_type"      : "linear",
+            "max_nodes_in_filter_radius": 100000,
+            "filter_radius_settings":{
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/filtering/explicit_vertex_morphing.py:21]: thank you!

File: applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp [hunk 4]
```diff
@@ -27,19 +27,20 @@ FilterFunction::FilterFunction(const std::string& rKernelFunctionType)
 
     if (rKernelFunctionType == "gaussian") {
         // Type 1: Gaussian function
-        mFilterFunctional =  [](double radius, double distance) {return std::max(0.0, exp(-(distance*distance)));};
+        // at distance = radius, the filter value will be 1e-8
+        mFilterFunctional =  [](double radius, double distance) { return std::max(0.0, exp(-(8 * 2.3025850929940455 * distance * distance) / (radius * radius))); };
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp]: where does this come from 2.3025850929940455  ?

File: applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp [hunk 5]
```diff
@@ -27,19 +27,20 @@ FilterFunction::FilterFunction(const std::string& rKernelFunctionType)
...
     } else if (rKernelFunctionType == "linear") {
         // Type 2: Linear function
-        mFilterFunctional =  [](double radius, double distance) {return std::max(0.0, (radius - distance) / radius);};
+        mFilterFunctional =  [](double radius, double distance) { return std::max(0.0, (radius - distance) / radius); };
     } else if (rKernelFunctionType == "constant") {
         // Type 3: Constant function
-        mFilterFunctional = [](double radius, double distance) {return 1.0;};
+        mFilterFunctional = [](double radius, double distance) { return 1.0; };
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp]:
```suggestion
        mFilterFunctional = [](double radius, double distance) { return (distance < radius) ? 1.0: 0.0 };
```

**Review by Igarizza — COMMENTED** [review]
nice! good work, thnx!

File: applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp (hunk 4 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/filtering/filter_function.cpp]: it is log_e(10)

File: applications/OptimizationApplication/tests/test_OptimizationApplication.py (hunk 1 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/tests/test_OptimizationApplication.py]: done :)

**Review by Igarizza — DISMISSED** [review]
good to go

**Review by RezaNajian — CHANGES_REQUESTED** [review]
Hi Suneth,
Since there are major changes, some time needed to review all changes.
Tnx.

**Comment by sunethwarna** [comment]
Closing in favor of #12201

**Pull request closed without merging**